File: LeetCodeCPlusPlus/py40.cpp
// 组合总数II，数字和等于target，不可重复；
// 思路：回溯，排序，同一层级里不允许使用两个相同的数
#include<vector>
using namespace std;
class Solution {
private:
    vector<int> tmp;
    int n;
    void helper(vector<int>& candidates, int index, int target, vector<vector<int>>& result) {
        if (target == 0) {
            result.push_back(tmp);
        }else {
            for (int i=index;i<n;i++) {
                if (i>index && candidates[i] == candidates[i-1])  // 同一层级的递归里不允许选择相同的项
                    continue;
                if (candidates[i] <= target) {
                    tmp.push_back(candidates[i]);
                    helper(candidates, i+1, target-candidates[i], result);
                    tmp.pop_back();
                }else
                    break;
            }
        }
    }
    
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        sort(candidates.begin(), candidates.end());
        n = candidates.size();
        helper(candidates, 0, target, result);
        return result;
    }
};

File: LeetCodePython/py40.py
# 组合总数II， 数字不可重复，使数字和为给定值
# 回溯
from typing import List
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        n = len(candidates)
        filter_set = set()
        tmp = []
        tmp_str = []
        def helper(pos, target):
            if target == 0:
                tmp_str = [str(item) for item in tmp]
                st = ",".join(tmp_str)
                if not st in filter_set:
                    result.append(list(tmp))
                    filter_set.add(st)
            elif pos>=n or candidates[pos]>target:
                return
            else:
                helper(pos+1, target)
                tmp.append(candidates[pos])
                helper(pos+1, target-candidates[pos])
                tmp.pop()
        candidates.sort()
        helper(0, target)
        return result
            